File: frontend/src/lib/api/billing.ts
import apiClient from './client';

export interface ProformaCompany {
  id?: number;
  nome: string;
  nif: string;
  morada: string;
  pais: string;
}

export interface ProformaItem {
  id: number;
  sku_marketplace?: string;
  internal_sku?: string;
  quantity: number;
  unit_price: number;
  vat_rate: number;
  vat_type?: string;
  vat_amount: number;
  line_total?: number;
  line_net?: number;
}

export interface ProformaOrder {
  id: number;
  empresa_id: number;
  external_order_id: string;
  marketplace_id?: number;
  order_date?: string;
  customer_country?: string;
  customer_name?: string | null;
  customer_address?: string | null;
  customer_nif?: string | null;
  currency?: string;
  total_gross: number;
  total_commission_fixed?: number;
  total_commission_percent?: number;
  total_net_value?: number;
  marketplace_nome?: string;
  items: ProformaItem[];
  total_vat?: number;
  total_net?: number;
  billing_document_id?: number | null;
  document_number?: string | null;
  document_status?: string | null;
}

export interface ProformaData {
  company: ProformaCompany;
  order: ProformaOrder;
}

export interface BillingDocumentItem {
  id: number;
  empresa_id?: number;
  sales_order_id: number;
  doc_type: string;
  document_number: string;
  status: string;
  total_gross?: number;
  total_net?: number;
  total_vat?: number;
  customer_country?: string;
  issued_at?: string;
  created_at?: string;
  cancelled_at?: string;
  external_order_id?: string;
  order_date?: string;
  marketplace_nome?: string;
}

export interface DocumentsListResponse {
  items: BillingDocumentItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface SimulationDocumentItem {
  tipo_documento: string;
  referencia_encomenda: string;
  valor_base: number;
  iva: number;
  total: number;
}

export interface SimulationResponse {
  items: SimulationDocumentItem[];
  total_faturas: number;
  total_nc: number;
  saldo_liquido: number;
}

export interface DocumentLine {
  artigo: string;
  quantidade: number;
  preco_unitario: number;
  taxa_iva: number;
  valor_iva: number;
  total_linha: number;
}

export interface DocumentPreview {
  tipo_documento: string;
  referencia_encomenda: string;
  cliente?: string;
  marketplace_nome?: string;
  valor_base: number;
  iva: number;
  total: number;
  linhas: DocumentLine[];
  sales_order_id?: number;
}

export interface VatSummaryItem {
  taxa_iva: number;
  base_tributavel: number;
  valor_iva: number;
}

export interface SimulationDetailedResponse {
  documentos: DocumentPreview[];
  total_faturas: number;
  total_nc: number;
  saldo_liquido: number;
  resumo_iva: VatSummaryItem[];
}

export const billingApi = {
  getProformaData(salesOrderId: number): Promise<ProformaData> {
    return apiClient.get(`/api/v1/billing/proforma-data/${salesOrderId}`).then((r) => r.data);
  },

  createProforma(salesOrderId: number): Promise<ProformaData> {
    return apiClient.post('/api/v1/billing/proformas', { sales_order_id: salesOrderId }).then((r) => r.data);
  },

  listDocuments(params?: {
    empresa_id?: number;
    doc_type?: string;
    status?: string;
    limit?: number;
    offset?: number;
  }): Promise<DocumentsListResponse> {
    return apiClient.get('/api/v1/billing/documents', { params }).then((r) => r.data);
  },

  cancelDocument(documentId: number): Promise<{ success: boolean; document_id: number }> {
    return apiClient.put(`/api/v1/billing/documents/${documentId}/cancel`).then((r) => r.data);
  },

  bulkCreateProformas(salesOrderIds: number[]): Promise<{
    created: Array<{ sales_order_id: number; document_number: string }>;
    skipped: number[];
    errors: number[];
  }> {
    return apiClient.post('/api/v1/billing/bulk-proformas', { sales_order_ids: salesOrderIds }).then((r) => r.data);
  },

  /** Processamento de Faturação Mensal (Batch Invoicing) */
  simulateBatch(params: {
    date_from: string;
    date_to: string;
    empresa_id?: number;
    marketplace_id?: number;
    serie_faturas?: string;
    serie_nc?: string;
  }): Promise<SimulationResponse> {
    return apiClient.post('/api/v1/billing/simulate', params).then((r) => r.data);
  },

  executeBatch(params: {
    date_from: string;
    date_to: string;
    empresa_id?: number;
    marketplace_id?: number;
    serie_faturas?: string;
    serie_nc?: string;
  }): Promise<{ success: boolean; message: string; created_faturas: number; created_nc: number }> {
    return apiClient.post('/api/v1/billing/execute', params).then((r) => r.data);
  },

  simulateBatchDetailed(params: {
    date_from: string;
    date_to: string;
    empresa_id?: number;
    marketplace_id?: number;
    serie_faturas?: string;
    serie_nc?: string;
  }): Promise<SimulationDetailedResponse> {
    return apiClient.post('/api/v1/billing/simulate-detailed', params).then((r) => r.data);
  },
};

File: backend/app/api/v1/billing.py
"""API Gestão Comercial: proformas, documentos de venda e processamento de faturação mensal (batch)."""
from fastapi import APIRouter, Depends, Query, HTTPException, Body
from typing import List, Optional, Any, Dict
from pydantic import BaseModel

from app.services.billing_service import BillingService
from app.api.deps import get_billing_service


router = APIRouter(prefix="/billing", tags=["billing"])


# ── Batch Invoicing (Processamento de Faturação Mensal) ─────────────────────

class SimulationRequest(BaseModel):
    """Parâmetros para simulação e execução do batch de faturação."""
    date_from: str
    date_to: str
    empresa_id: Optional[int] = None
    marketplace_id: Optional[int] = None
    serie_faturas: Optional[str] = None
    serie_nc: Optional[str] = None


class SimulationDocumentItem(BaseModel):
    tipo_documento: str
    referencia_encomenda: str
    valor_base: float
    iva: float
    total: float


class SimulationResponse(BaseModel):
    items: List[SimulationDocumentItem]
    total_faturas: float
    total_nc: float
    saldo_liquido: float


# ── Simulate detalhado (master-detail + resumo IVA) ─────────────────────────

class DocumentLineResponse(BaseModel):
    artigo: str
    quantidade: float
    preco_unitario: float
    taxa_iva: float
    valor_iva: float
    total_linha: float


class DocumentPreviewResponse(BaseModel):
    tipo_documento: str
    referencia_encomenda: str
    cliente: Optional[str] = None
    marketplace_nome: Optional[str] = None
    valor_base: float
    iva: float
    total: float
    linhas: List[DocumentLineResponse] = []
    sales_order_id: Optional[int] = None


class VatSummaryItemResponse(BaseModel):
    taxa_iva: float
    base_tributavel: float
    valor_iva: float


class SimulationDetailedResponse(BaseModel):
    documentos: List[DocumentPreviewResponse]
    total_faturas: float
    total_nc: float
    saldo_liquido: float
    resumo_iva: List[VatSummaryItemResponse]


@router.post("/simulate", response_model=SimulationResponse)
async def simulate_batch_invoicing(
    body: SimulationRequest,
    service: BillingService = Depends(get_billing_service),
):
    """
    Simula o processamento de faturação mensal.
    Devolve encomendas elegíveis para Fatura e devoluções para NC, com totais.
    Não altera o estado na BD (idempotente e seguro).
    """
    try:
        result = service.simulate_batch(
            date_from=body.date_from,
            date_to=body.date_to,
            empresa_id=body.empresa_id,
            marketplace_id=body.marketplace_id,
            serie_faturas=body.serie_faturas,
            serie_nc=body.serie_nc,
        )
        items = [
            SimulationDocumentItem(
                tipo_documento=it["tipo_documento"],
                referencia_encomenda=it["referencia_encomenda"],
                valor_base=it["valor_base"],
                iva=it["iva"],
                total=it["total"],
            )
            for it in result["items"]
        ]
        return SimulationResponse(
            items=items,
            total_faturas=result["total_faturas"],
            total_nc=result["total_nc"],
            saldo_liquido=result["saldo_liquido"],
        )
    finally:
        service.close()


@router.post("/simulate-detailed", response_model=SimulationDetailedResponse)
async def simulate_batch_detailed(
    body: SimulationRequest,
    service: BillingService = Depends(get_billing_service),
):
    """
    Simula faturação com detalhe por linha (master-detail) e resumo IVA por taxa.
    Totais = soma exata das linhas.
    """
    try:
        result = service.simulate_batch_detailed(
            date_from=body.date_from,
            date_to=body.date_to,
            empresa_id=body.empresa_id,
            marketplace_id=body.marketplace_id,
            serie_faturas=body.serie_faturas,
            serie_nc=body.serie_nc,
        )
        docs = [
            DocumentPreviewResponse(
                tipo_documento=d["tipo_documento"],
                referencia_encomenda=d["referencia_encomenda"],
                cliente=d.get("cliente"),
                marketplace_nome=d.get("marketplace_nome"),
                valor_base=d["valor_base"],
                iva=d["iva"],
                total=d["total"],
                linhas=[
                    DocumentLineResponse(
                        artigo=ln["artigo"],
                        quantidade=ln["quantidade"],
                        preco_unitario=ln["preco_unitario"],
                        taxa_iva=ln["taxa_iva"],
                        valor_iva=ln["valor_iva"],
                        total_linha=ln["total_linha"],
                    )
                    for ln in d.get("linhas", [])
                ],
                sales_order_id=d.get("sales_order_id"),
            )
            for d in result["documentos"]
        ]
        return SimulationDetailedResponse(
            documentos=docs,
            total_faturas=result["total_faturas"],
            total_nc=result["total_nc"],
            saldo_liquido=result["saldo_liquido"],
            resumo_iva=[
                VatSummaryItemResponse(
                    taxa_iva=x["taxa_iva"],
                    base_tributavel=x["base_tributavel"],
                    valor_iva=x["valor_iva"],
                )
                for x in result["resumo_iva"]
            ],
        )
    finally:
        service.close()


@router.post("/execute")
async def execute_batch_invoicing(
    body: SimulationRequest,
    service: BillingService = Depends(get_billing_service),
) -> Dict[str, Any]:
    """
    Executa a faturação definitiva: cria Faturas e Notas de Crédito,
    atribui números sequenciais e marca as encomendas como faturadas.
    Ação irreversível.
    """
    try:
        result = service.execute_batch(
            date_from=body.date_from,
            date_to=body.date_to,
            empresa_id=body.empresa_id,
            marketplace_id=body.marketplace_id,
            serie_faturas=body.serie_faturas,
            serie_nc=body.serie_nc,
        )
        return result
    finally:
        service.close()


class ProformaDataResponse(BaseModel):
    company: dict
    order: dict


class BulkProformasRequest(BaseModel):
    sales_order_ids: List[int]


class DocumentListItem(BaseModel):
    id: int
    empresa_id: Optional[int]
    sales_order_id: int
    doc_type: str
    document_number: str
    status: str
    total_gross: Optional[float]
    total_net: Optional[float]
    total_vat: Optional[float]
    customer_country: Optional[str]
    issued_at: Optional[str]
    created_at: Optional[str]
    cancelled_at: Optional[str]
    external_order_id: Optional[str]
    order_date: Optional[str]
    marketplace_nome: Optional[str]


class DocumentsListResponse(BaseModel):
    items: List[DocumentListItem]
    total: int
    limit: int
    offset: int


@router.get("/proforma-data/{sales_order_id}", response_model=ProformaDataResponse)
async def get_proforma_data(
    sales_order_id: int,
    service: BillingService = Depends(get_billing_service),
):
    """Dados para preview/impressão da proforma (empresa, ordem, linhas com IVA)."""
    try:
        data = service.get_proforma_data(sales_order_id)
        if not data:
            raise HTTPException(status_code=404, detail="Venda não encontrada")
        return ProformaDataResponse(company=data["company"], order=data["order"])
    finally:
        service.close()


@router.post("/proformas", response_model=ProformaDataResponse)
async def create_proforma(
    sales_order_id: int = Body(..., embed=True),
    service: BillingService = Depends(get_billing_service),
):
    """Cria proforma para uma venda e devolve dados para impressão."""
    try:
        data = service.create_proforma(sales_order_id)
        if not data:
            raise HTTPException(status_code=404, detail="Venda não encontrada")
        return ProformaDataResponse(company=data["company"], order=data["order"])
    finally:
        service.close()


@router.get("/documents", response_model=DocumentsListResponse)
async def list_documents(
    empresa_id: Optional[int] = Query(None),
    doc_type: Optional[str] = Query(None, description="Proforma, Fatura, NC"),
    status: Optional[str] = Query(None, description="issued, cancelled"),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    service: BillingService = Depends(get_billing_service),
):
    """Lista documentos comerciais com filtros."""
    try:
        items, total = service.list_documents(
            empresa_id=empresa_id,
            doc_type=doc_type,
            status=status,
            limit=limit,
            offset=offset,
        )
        out = [
            DocumentListItem(
                id=r["id"],
                empresa_id=r.get("empresa_id"),
                sales_order_id=r["sales_order_id"],
                doc_type=r["doc_type"],
                document_number=r["document_number"],
                status=r["status"],
                total_gross=r.get("total_gross"),
                total_net=r.get("total_net"),
                total_vat=r.get("total_vat"),
                customer_country=r.get("customer_country"),
                issued_at=str(r["issued_at"]) if r.get("issued_at") else None,
                created_at=str(r["created_at"]) if r.get("created_at") else None,
                cancelled_at=str(r["cancelled_at"]) if r.get("cancelled_at") else None,
                external_order_id=r.get("external_order_id"),
                order_date=str(r["order_date"]) if r.get("order_date") else None,
                marketplace_nome=r.get("marketplace_nome"),
            )
            for r in items
        ]
        return DocumentsListResponse(items=out, total=total, limit=limit, offset=offset)
    finally:
        service.close()


@router.put("/documents/{document_id}/cancel")
async def cancel_document(
    document_id: int,
    service: BillingService = Depends(get_billing_service),
):
    """Anula um documento comercial."""
    try:
        ok = service.cancel_document(document_id)
        if not ok:
            raise HTTPException(status_code=404, detail="Documento não encontrado")
        return {"success": True, "document_id": document_id}
    finally:
        service.close()


@router.post("/bulk-proformas")
async def bulk_create_proformas(
    body: BulkProformasRequest,
    service: BillingService = Depends(get_billing_service),
):
    """Cria proformas em lote para várias vendas. Ignora as que já têm proforma."""
    try:
        result = service.bulk_create_proformas(body.sales_order_ids)
        return result
    finally:
        service.close()
